TEPHRA-228 Adding the ability to pass-in a clientId during the start …

gokulavasan opened the pull request:
…of a transaction which is logged when the transaction gets invalidated during time out.

gokulavasan commented on tephra-core/src/main/java/org/apache/tephra/TransactionManager.java:
```diff
@@ -763,15 +781,22 @@ private long getNextWritePointer() {
+  /**
+   * Starts a long transaction with a client id.
+   */
+  public Transaction startLong(String clientId) {
```
If we are going to deprecate the clientId-less methods, should we allow null here?

gokulavasan replied:
Should I add Nullable annotation if we decide to allow null for clientId?

poornachandra replied:
Eventually we want all client's to pass in a client i, hence the deprecation. If we allow nulls then that defeats the purpose, no?

gokulavasan replied:
@poornachandra Sure sounds good. I will add a Precondition check for String not being null.

gokulavasan replied:
Added Preconditions null check for clientId.

gokulavasan commented on tephra-core/src/main/java/org/apache/tephra/TransactionManager.java:
```diff
@@ -1378,17 +1403,32 @@ public TransactionType getTransactionType() {
     private final long expiration;
     private final InProgressType type;
     private final LongArrayList checkpointWritePointers;
+    private final String clientId;
```
Not sure if we can add it to equals/hashCode. The unit tests fail when trying to compare the state before persistence and after recovery and won't be the same if we include clientId in comparison. Any reason why we should definitely include it? Or should we include it and adjust the test that does the comparison?

poornachandra replied:
It is okay not to add it now since we are not persisting it. Add a comment in the code.

gokulavasan commented:
@poornachandra Addressed comments. Please take a look.

gokulavasan commented on tephra-core/src/main/java/org/apache/tephra/TransactionSystemClient.java:
```diff
@@ -44,11 +44,33 @@
   Transaction startShort(int timeout);
```
Done.

poornachandra commented on tephra-hbase-compat-0.98/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java:
```diff
@@ -127,7 +126,7 @@ public TransactionProcessor() {
     if (e instanceof RegionCoprocessorEnvironment) {
       RegionCoprocessorEnvironment env = (RegionCoprocessorEnvironment) e;
-      Supplier<TransactionStateCache> cacheSupplier = getTransactionStateCacheSupplier(env);
+      this.cacheSupplier = getTransactionStateCacheSupplier(env);
```
Unrelated change?

gokulavasan replied:
Missed this change in https://github.com/apache/incubator-tephra/pull/41 for this compat-module. Hence fixing it here.

gokulavasan commented:
@poornachandra Can you please review the last commit. I have addressed your comment about TransactionContext change (deprecated old constructors and introduced a new one with the clientId)

gokulavasan commented:
@poornachandra Injecting ClientId in TransactionServiceClient now instead of changing the API of TransactionSystemClient. Please review when you get a chance.

gokulavasan commented:
@poornachandra Please review the last commit. I have addressed your comment about including hostname-pid as the default client id.

poornachandra commented on tephra-core/src/main/java/org/apache/tephra/runtime/ClientIdAwareTransactionModule.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Guice module that binds the clientId to be used in {@link TransactionSystemClient}.
+ */
+public abstract class ClientIdAwareTransactionModule extends AbstractModule {
```
Having `ClientIdAwareTransactionModule` as the parent class of Transaction modules does not add much value. Let's move the binding to the child modules.

gokulavasan replied:
Sure ok.

poornachandra commented on tephra-core/src/main/java/org/apache/tephra/runtime/TransactionDistributedModule.java:
```diff
@@ -35,13 +34,24 @@
+final class TransactionDistributedModule extends ClientIdAwareTransactionModule {
+
+  public TransactionDistributedModule() {
+    super(ManagementFactory.getRuntimeMXBean().getName());
```
This default is already defined in `TransactionModules`. Do we still need it here? 
Same for `TransactionInMemoryModule` and `TransactionLocalModule`.

gokulavasan replied:
@poornachandra There is no guarantee that TransactionModules().getDistributedModule() is used to get an instance of TransactionDistributedModule. One can just create an instance of this class.

poornachandra replied:
In that case they can provide the client id since the default constructor won't be present, right?

gokulavasan commented:
@poornachandra Thanks for the comments. Addressed them.

poornachandra commented on tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java:
```diff
@@ -130,10 +134,11 @@ public static void doMain(boolean verbose, Configuration conf) throws Exception
   @Inject
-  public TransactionServiceClient(Configuration config,
-                                  ThriftClientProvider clientProvider) {
+  public TransactionServiceClient(Configuration config, ThriftClientProvider clientProvider,
```
Some clients like Apache Phoenix do not use the guice modules. They instantiate `TransactionServiceClient` directly. It will be good to have the existing constructor too and assign the default value for the clientId in it.

gokulavasan replied:
@poornachandra Addressed comment in the commit https://github.com/apache/incubator-tephra/pull/42/commits/0279658b3b0718298ccfe363343bce182e4ba655. Please take a look again.

poornachandra commented on tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java:
```diff
@@ -125,15 +130,20 @@ public static void doMain(boolean verbose, Configuration conf) throws Exception
     }
   }
 
+  public TransactionServiceClient(Configuration config, ThriftClientProvider clientProvider) {
```
Please add javadocs for the public constructor.
